docs/developer-guide/writing-actions-reducers.md

```diff
@@ -0,0 +1,203 @@
+                    })
+                };
+            }
+            default return state;
+        }
+    }
+);
+```
+Which is good but maybe less readable than using other ways.
+
+We suggest to use lodash/fp which allows to avoid mutation of the objects (remember reducers are pure function)
+We have prepared some utilities in @mapstore/utils/ImmutableUtils like set
+
+The previous example with set() becomes:
+```
+import {findIndex} from 'lodash';
+...
+    case UPDATE_LAYER_FEATURE: {
+        const indexFeatureChanged = findIndex(state.features, id => action.id)
+        const features = indexFeatureChanged !== -1 ? set([indexFeatureChanged], action.props, state.features) : state.features;
+        return {
+            ...state,
+            features
+        };
+    }
+...
+```
+
+# Testing
+Tests in mapstore are stored in `__tests__` folder at the same level where actions/reducer are.
+The file name is the same of the action/reducer with a '-test' suffix
+```
+actions/map.js
+actions/__tests__/map-test.js
+or
+reducers/map.js
+reducers/__tests__/map-test.js
+```
+
+We use [expect](https://github.com/mjackson/expect) as testing library, therefore we suggest to have a look there.
+
+## How to test and action
+Typically you want to test the type and the params return from the action creator
+
+let's test the mapTo action:
+```
+// copyright section
+import expect from 'expect';
+import {panTo, PAN_TO} from '@mapstore/actions/map';
+describe('Test correctness of the map actions', () => {
+    it('testing panTo', () => {
+        const center = [2, 3];
+        const returnValue = panTo(center);
+        expect(returnValue.type).toEqual(PAN_TO);
+        expect(returnValue.center).toEqual(center);
+    });
+});
+
+```
+In order to speed up the unit test runner, you can:
+- change the path in build\tests.webpack.js to point to the folder parent of __tests__ for example '../web/client/actions'
+- then run this command:
+`npm run continuoustest`
+
+This allows to run only the tests contained to the specified path.
+**Note:** When all tests are successfully passing remember to restore it to its orignal value '../web'
+
+## How to test an reducer
+Here things can become more complicated depending on your reducer but in general you want to test all cases
+```
+// copyright section
+import expect from 'expect';
+import {panTo} from '@mapstore/actions/map';
+import map from '@mapstore/reducers/map';
+describe('Test correctness of the map reducers', () => {
+    it('testing PAN_TO', () => {
+        const center = [2, 3];
+        const state = map({}, panTo(center));
+
+        // here you have to check that state has changed accordingly
+        expect(state.center).toEqual(center);
+    });
+});
+```
+
+Here for speedup testing you can again modify the build\tests.webpack.js
+in order to point to the reducers folder and then runnning
+`npm run continuoustest`
+
+## Actions and epics
```

Comment:
This paragraph is not needed at all

Reply:
why? i'm just mentioning that actions (constants) are bound to epics, so if anyone starts reading from this page can jump to the epics doc directly without having to navigate the documentation. 

i would not remove it
@mbarto @tdipisa  opinions?

Reply:
If you want to mention epics, I would say something more meaningful like "Actions are not only used by redux to update the store (through the reducers) but also for triggering side effects workflows managed by epics".